### ISSUE TITLE
Autocomplete - updating bound SelectedValue from codebehind won't update the SelectedText

### ISSUE THREAD

**Issue by MichalCadecky** [Megabit/Blazorise#2879]
**Describe the bug**
I believe this has been broken by the newest improvements to *Autocomplete* in #2548. The whole `SelectedText` parameter introduction is quite a breaking change which was **NOT** documented very well unfortunately.

Currently if you bind the `SelectedValue` to a codebehind property, updating it will not update the newly introduced `SelectedText` as well. This is noticable especially in *edit*-like forms where you need to initialize *Autocomplete* with initial value but also when you have a presentation logic causing the `SelectedValue` to change (e.g. to `null`).

Changing the `SelectedText` parameter will update the `SelectedValue` properly - thanks to the `OnTextChangedHandler` implementation.

**Expected behavior**
I expected the `SelectedValue` to work as it was working before the improvements introduced in 0.9.4, if I am not intrested in managing `SelectedText` by myself. Now I have to manage state of two properties even though I have already told the component where it can get the text to display.

There seems to be already something prepared (which is actually currently only wasting performance) in the setter implementation of `SelectedValue` (lines 423-425).

https://github.com/Megabit/Blazorise/blob/8c5deda7ab61279036865e54a085e786c05063d4/Source/Extensions/Blazorise.Components/Autocomplete.razor.cs#L412-L427

I understand that updating the `SelectedText` from the setter of `SelectedValue` can possibly cause infinite loop of event callbacks. I am not really a fan of the introduced `SelectedText` functionality because it caused a lot of confusion in my opinion. There is still the `CurrentSearch` functionality as well and the respective events `SelectedTextChanged` and `SearchChanged` are almost completely duplicate. 

I am not sure what is the correct way out of this. I have seen somewhere that Blazorise team is going to completely remove the value selection behavior in *Autocomplete* and instead introduce autocomplete-like filtering in *Select* components in the future. But for now, this is really a hybrid that is not working very well.

**Comment by David-Moreira** [comment]
Hello @MichalCadecky 
First of all, we are sorry if something is now not working as it was before. It was not intended to do a breaking change in this component and we always do make a point to inform our users of breaking changes on our release notes, so with this change, we only introduced opt in features which fight some of the limitations of the Autocomplete. We'd like it to be feature rich, giving even more options to our users, features that they'd expect from a component like this.

As such everything should be working as it was before, just with the option to activate / use the additional features.

Could you clarify what is not working for you? You should not need to use or manage `SelectedText` if you weren't before. And `SelectedValue` should be working as it was before, which is to bind to a value that exists in the data source collection.

Now `SelectedText` is new, and solves a different problem. I might want the component to just aid my user with suggestions, but still want to track the value that he has typed. So `SelectedText` / `FreeTypying` comes into play, tracking that value which is outside of the Data Source's values and allowing validation to take effect in `SelectedText` instead. So these were never intended to have to be managed "together", unless your use case requires to do so.


`CurrentSearch` and `SelectedText` do seem similar. But the code isn't 100% the same for both. I can't recall why we didn't "re-use" `CurrentSearch` at the moment, but I'm sure there was a reason behind the new property, but we can investigate this matter and make sure we didn't introduce a duplicate api unnecessarily. I will get back to you on this one.
@stsrki by any chance, do you remember?


Would appreciate an example, and if you can explain the **Before** / **After** behaviour, that would also be much appreciated! Thanks!

**Comment by David-Moreira** [comment]
Hello @MichalCadecky 
So I am now working on a new feature for `Autocomplete` and took a look at this. The reason why there is `SelectedText` and `CurrentSearch`, is because `CurrentSearch` literally takes in the string being searched by the user, while `SelectedText` does the same, it also goes a bit further, listening for the `TextField` Parameter that can mutate the display / text according to the selected item, which means, the `SelectedText` might not be always the same as `CurrentSearch` as is "faithfull" to the `TextField` Parameter.

![image](https://user-images.githubusercontent.com/22283161/135732344-e53b8c18-d9ef-48fd-b92b-9bc0a87e349b.png)

I think it's safe to say both are needed, and you can choose to manage or not these according to your use case.
Just let us know if you have any more questions or something you find unusual.

**Comment by MichalCadecky** [comment]
Hi @David-Moreira!

Sorry for delayed response. First of all thanks for the great job you are doing. It was not my intention to acuse you or put a guilt on you in any way. I just wanted to inform you about the situation we are currently experiencing.

You have focused on my observation of how `CurrentSearch` and `SelectedText` are influencing each other, but the main issue is between `SelectedText` and `SelectedValue` as I have already described.

Setting `SelectedValue` from codebehind will not cause the `SelectedText` to update. That is why I have mentioned the need of managing `SelectedText` independently. Before the introduction of `SelectedText` property updating the `SelectedValue` would result in rendering of respective options' text. This is not the case anymore unfortunately.

**Comment by David-Moreira** [comment]
Hello,
It seems like I did not understand the issue at hand. You should not need to manage it, no. It should be managed for you. Seems like a bug, also just reported on #2991 / #2993 
Sorry about the incovenience!!

**Comment by MichalCadecky** [comment]
Indeed, those look like complete duplicate of this issue.

I have made quite intensive debugging session to understand what, where and when is updating. There are things happening in parallel and it is currently very hard to debug this kind of parallelism in Blazor.

Currently two events are very relevant. `OnTextChangedHandler` and `OnTextKeyDownHandler` are happening in parallel but these does not seem to cause issues. There is a 3rd event happening somewhere but I am not able to trace it as the call stack is always showing the line where the evencallback with long-running tasks is being invoked. I thought it might be either the internal events of `TextEdit` or something relevant to the re-rendering of the component. But that is just my theory. This 3rd event is accessing and setting the `SelectedText` property also causing very strange behavior. 

Also long-running tasks in eventcallbacks like `SearchChanged` (e.g. this is being done in ABP lookups) are also currently impacted by the respective changes. If the long-running task is being worked on and Autocomplete gets re-rendered during this time, `NotFoundVisible` will be `true` and the `NotFoundContent` will be shown even though the `NotFound` event callback will never be invoked.

**Comment by SethVanderZanden** [comment]
Hey guys just writing a follow up as I could be misunderstanding. Was the solution that the SelectedText would automatically be updated if the selectedValue was selected in codebehind? I had an issue for awhile and my only fix was to specify the SelectedText otherwise it would remain blank, even tho I had updated the search text and selected value.

`    if (_newCatch.FarmId != 0)
            {
                _selectedFarm = _farmSelectList.Data.First(f => f.FarmId == _newCatch.FarmId);
                _farmSearchText = _selectedFarm.FarmName;
            }`

Called from a function inside the OnInitAsync. The only fix was to specify the selected text, if I did not it didn't show the selected items. Is this intentional or a bug? In the past if I had set the Search Text and Selected Value, it worked fine. However, since making it pre-load the selected value from my OnInitAsync, it has since caused the addition of the SelectedText

`   <Autocomplete
                            TItem="ProcessorSelectDto"
                            TValue="ProcessorSelectDto"
                            Data="@_processorSearchResults"
                            TextField="@((item) => item.ProcessorName)"
                            ValueField="@((item) => item)"
                            ReadData="(async (e) => await OnFilterProcessors(e))"
                            Placeholder="Search Processors"
                            Filter="AutocompleteFilter.StartsWith"
                            SearchChanged="@(async (e) => await SetProcessorSearch(e))"
                            SelectedValueChanged="@(async (e) => await SetProcessor(e))"
                            SelectedText="@_selectedProcessor?.ProcessorName"
                            Class="form-control-solid w-100"
                            SearchClass="form-control-solid"
                            Context="processorAutoCompleteContext"
                            AutoPreSelect=true>`
